### DISCUSSION TITLE
Finding Vertical Lines (signals) in PNGs and getting its x position

### DISCUSSION THREAD

**Discussion by NoRainDropsInTheSky** [ImageMagick/ImageMagick#2782] [Help] [answered]
Hello. 

I am working on finding signals (basically thick vertical lines against a static background (see example below) in PNGs. This is what I have so far:

Original PNG with signal (vertical line)
![400MHz_1](https://user-images.githubusercontent.com/73752919/97784644-ac391100-1b65-11eb-8321-fb0430e7cdbe.png)

**Step 1**:  Apply an 80% thresholding to get rid of the background noise and show only the signal
`convert signal.png -threshold 80% -depth 1 signal_2.png`

![400MHz_2](https://user-images.githubusercontent.com/73752919/97784653-b78c3c80-1b65-11eb-86f7-fedb78f8fdbf.png)

**Step 2**:  Find edges
`convert signal_2.png -canny 0x1+10%+30% signal_3.png`

![400MHz_3](https://user-images.githubusercontent.com/73752919/97784717-1782e300-1b66-11eb-83ed-b37671471882.png)

**Step 3**: Apply 'hough line algorithm' to show the vertical line pattern as a red line
`convert signal_3.png -background black -stroke red -hough-lines 20x20+120 signal_4.png
`
![400MHz_4](https://user-images.githubusercontent.com/73752919/97784741-4436fa80-1b66-11eb-9080-8323b0179d9c.png)

**Step 4**: Find the position (x-coordinate) of the red line (hence the position of the signal) by converting PNG into {x,y} colormap and find red color (color 36863)
`convert signal_4.png txt:- | grep "36863" | head -n600 | tail -n1 | awk -F, '{print $1}' > signal_position.txt`

`363`

Thus, the signal is located at pixel number 363 (along the x axis)

QUESTIONS:  

1. **Is there a better, more sure-proof way of going about this**  ? If there are background 'noise' (aka other diagonal lines or horizontal lines, the thresholding doesn't get rid of them)
2. The 'hough line algorithm' generates mutliple red lines...and when you try to find them using grep from the colormap text file, it finds multiple x locations...**is it possible to get just one red line where the signal is** ?

**Comment by fmw42** [comment]
You can try this. Use some morphology to clean up the image. Convert to gray, use otsu automatic thresholding, scale to one row and threshold again so that you have black and white. Find all the white pixels.

```
magick line.png -alpha off \
-morphology close:9 square:1 \
-colorspace gray \
-auto-threshold otsu \
+write threshold.png \
-scale x1! \
-threshold 0 \
txt: | grep "gray(255)"
```

![threshold](https://user-images.githubusercontent.com/13458218/97785734-181a7a00-1b64-11eb-9968-4a93a328a935.png)



```
105,0: (255)  #FFFFFF  gray(255)
106,0: (255)  #FFFFFF  gray(255)
107,0: (255)  #FFFFFF  gray(255)
108,0: (255)  #FFFFFF  gray(255)
109,0: (255)  #FFFFFF  gray(255)
110,0: (255)  #FFFFFF  gray(255)
111,0: (255)  #FFFFFF  gray(255)
112,0: (255)  #FFFFFF  gray(255)
```

Get the first and last or center x coordinate and then draw a line

`magick line.png -fill red -draw "line 112,0 105,%[fx:h-1]" line_red.png`

![line_red](https://user-images.githubusercontent.com/13458218/97785732-12249900-1b64-11eb-837a-a39e5e4b25ee.png)

**Comment by NoRainDropsInTheSky** [comment]
Hi **fmw42**, and thank you for responding so quickly.

I forgot to mention my base system:

`Linux Mint 20 Ulyana`

`Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP 
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib`

When I type in your command (with the PNG renamed to line.png), I get the error 

`bash: magick: command not found`

So instead I tried using the `convert` 

`convert-im6.q16: unrecognized option '-auto-threshold' @ error/convert.c/ConvertImageCommand/748.`

Then I tried to use `mogrify`

`mogrify-im6.q16: unrecognized option '-auto-threshold' @ error/mogrify.c/MogrifyImageCommand/4112`

**Comment by fmw42** [comment, marked as answer]
-auto-threshold is only in IM 7. So you will need to replace it with a simple -threshold XX%

(Or upgrade to IM 7)

```
convert line.png -alpha off \
-morphology close:9 square:1 \
-colorspace gray \
-threshold 65% \
+write threshold.png \
-scale x1! \
-threshold 0 \
txt: | grep "gray(255)"
```


![threshold](https://user-images.githubusercontent.com/13458218/97787590-76e5f080-1b70-11eb-821d-2e2c61b2ea57.png)

```
105,0: (255,255,255)  #FFFFFF  gray(255)
106,0: (255,255,255)  #FFFFFF  gray(255)
107,0: (255,255,255)  #FFFFFF  gray(255)
108,0: (255,255,255)  #FFFFFF  gray(255)
109,0: (255,255,255)  #FFFFFF  gray(255)
110,0: (255,255,255)  #FFFFFF  gray(255)
111,0: (255,255,255)  #FFFFFF  gray(255)
112,0: (255,255,255)  #FFFFFF  gray(255)

```

`convert line.png -fill red -draw "line 112,0 105,374" line_red.png`

![line_red](https://user-images.githubusercontent.com/13458218/97787641-aeed3380-1b70-11eb-851a-5b81a431e7f6.png)
